### ISSUE TITLE
Retrain model with user image

### ISSUE THREAD

**Issue by mathieuboudreau** [axondeepseg/axondeepseg#833]
### Discussed in https://github.com/ivadomed/ivadomed/discussions/1324

<div type='discussions-op-text'>

<sup>Originally posted by **MirenLurBarquin** October 10, 2024</sup>
Dear all, 

I am currently using [AxonDeepSeg](https://axondeepseg.readthedocs.io/en/latest/#) and [Ivadomed](https://ivadomed.org/tutorials/two_class_microscopy_seg_2d_unet.html) for my project; however, I have some questions as I am not obtaining the expected results. 

I have a TEM dataset of rat brain tissue, for which axon and myelin segmentation was performed using AxonDeepSeg. The images are in `.tif` format, of size `10.24 x 10.24 µm` and a resolution of `2048 x 2048 pixels` (i.e.  a pixel size of `0.005 x 0.005 µm`). 

Due to the characteristics of the image (e.g., contrast and definition) the segmentation outputs exhibit issues such as **holes, sharp edges, and missing axons and myelin**. To address this, we attempted to postprocess these segmentations, retaining only the blobs that appeared accurate in each mask. This approach resulted in a **sparse segmentation** dataset.

Next, we retrained the model using the sparse dataset, hoping to achieve a dense segmentation better than the one resulted from AxonDeepSeg. Unfortunately, this was not the outcome. For the retraining process, we utilized Ivadomed.

I have attached the latest configuration file I am using (having experienced with various other configurations), and I will outline some specific notes and questions that may be influencing the unsatisfactory results:  

1. By default Ivadomed uses `DiceLoss`, but after not obtaining a dense segmentation I tried to use the `GeneralizedDiceLoss` with no background. Which obtained much worse results than the base case: empty axon class for `"transfer_learning"{..., "retrain_fraction": 1.0, "reset": true}` and just small axons for `"transfer_learning"{..., "retrain_fraction": 0.3, "reset": true}`. **Is this idea right?: `no background` masks out the background from the calculation of the loss to help the model not to _penalize_ when new axons and myelin are found in the background.**

2. Has someone solved a similar task using Ivadomed and AxonDeepSeg?

Any idea or input is more than welcome :) 
Thanks in advance!

[config_microscopy_base.json](https://github.com/user-attachments/files/17328551/config_microscopy_base.json)

</div>

@hermancollin - I tried segmenting the user's image using our latest ADS model and it didn't perorm as well as IVADOMED,

![375424870-621a87ed-7757-4987-810f-c16b71369f0f](https://github.com/user-attachments/assets/3b16c6c4-836f-4ad7-b209-f1d555ac5947)

vs IVADOMED:

![375392947-6e482f49-8c90-4560-b199-69fe275e8621](https://github.com/user-attachments/assets/ec4f5b86-6f7d-4073-bec3-599ba9a12f8e)

**Comment by hermancollin** [comment]
To be honest, I'm quite impressed by the ivadomed model performance here. I guess the nnunet models have a tendency to overfit their training data a little bit.

The pixel size of this image is ~2x the px size of our TEM dataset. I tried resizing the image to make it twice as big, but it didn't really help... 

<details>

![mirenlurbarquin_gh_user_bigger_seg-axonmyelin](https://github.com/user-attachments/assets/a37fe137-4d9d-4a2e-a40d-c5e68e66e0bb)
</details>

The best result I was able to achieve was by splitting the image in 4 and equalizing the histograms for all quarters. The reason why I split the image in 4 is because there is a gradient in the image (top-right corner is bright, bottom-left is dark) and it gives better contrasts. However, the results are still no amazing, but at least it picks up some big axons:

<details>

![mirenlurbarquin_gh_user_bigger_ROI-0_seg-axonmyelin](https://github.com/user-attachments/assets/d303b809-4de1-4168-a6de-bfff2ce78dc7)
![mirenlurbarquin_gh_user_bigger_ROI-1_seg-axonmyelin](https://github.com/user-attachments/assets/5d303e4f-26d9-441c-9551-71deb27c4d71)
![mirenlurbarquin_gh_user_bigger_ROI-2_seg-axonmyelin](https://github.com/user-attachments/assets/dc2467e2-b717-4d06-843e-19f98d4381b7)
![mirenlurbarquin_gh_user_bigger_ROI-3_seg-axonmyelin](https://github.com/user-attachments/assets/9cb88a0f-0eda-4c20-84bd-fba9055767ca)


</details>

**Comment by jcohenadad** [comment]
> The fact that ivadomed would resize the images to match the training resolution was definitely advantageous in some cases, but we lost that with nnunet. Unfortunately, this doesn't really help here anyway.

should we consider re-training with nnUnet, but this time resampling the images to a fixed resolution?

**Comment by hermancollin** [comment]
> > The fact that ivadomed would resize the images to match the training resolution was definitely advantageous in some cases, but we lost that with nnunet. Unfortunately, this doesn't really help here anyway.
> 
> should we consider re-training with nnUnet, but this time resampling the images to a fixed resolution?

This would most likely degrade a lot of our masks. The resolutions of all our datasets span ~2 orders of magnitude.

It might be important to add back a resampling option in ADS (i.e. as a preprocessing step), but the users don't necessarily know the resolution of our training data so they couldn't match the resolution themselves. Something I have been thinking about for some time would be to train a resampling module to automatically resize the images.

**Comment by jcohenadad** [comment]
> Something I have been thinking about for some time would be to train a resampling module to automatically resize the images.

that sounds like a great idea.
